flake: TestServer/Logging fails with `no rows returned for version select`

Seen here: https://github.com/coder/coder/actions/runs/15461129890/job/43522552318

```
=== FAIL: cli TestServer/Logging/CreatesFile (2.82s)
    clitest.go:75: invoking command: coder --global-config /var/folders/y4/8dcllrc96x32z0w3fgrlx6_m0000gn/T/TestServerLoggingCreatesFile2486680519/002 server --log-filter=.* --postgres-url=***127.0.0.1:5432/test_2025_06_05_07_35_16_jxjvtgijva?sslmode=disable --http-address :0 --access-url http://example.com/ --provisioner-daemons=3 --provisioner-types=echo --log-human /var/folders/y4/8dcllrc96x32z0w3fgrlx6_m0000gn/T/TestServerLoggingCreatesFile2486680519/001/coder-logging-test-1008095851
    t.go:106: 2025-06-05 07:35:16.826 [debu]  cli: started debug logging
    t.go:106: 2025-06-05 07:35:16.826 [info]  cli: stdout: Started HTTP listener at http://[::]:51869
    t.go:106: 2025-06-05 07:35:16.826 [info]  cli: ...
        msg= stdout: ╔════════════════════════╗
             ║    View the Web UI:    ║
             ║   http://example.com/   ║
             ╚════════════════════════╝
    t.go:106: 2025-06-05 07:35:16.826 [debu]  cli: connecting to postgresql
    t.go:106: 2025-06-05 07:35:16.904 [debu]  cli: closing tracing
    t.go:106: 2025-06-05 07:35:16.904 [debu]  cli: tracing closed  error=<nil>
    clitest.go:261: command "coder server" exited with error: running command "coder server": connect to postgres:
            github.com/coder/coder/v2/cli.(*RootCmd).Server.func2
                /Users/runner/work/coder/coder/cli/server.go:754
          - connect to postgres:
            github.com/coder/coder/v2/cli.getAndMigratePostgresDB
                /Users/runner/work/coder/coder/cli/server.go:2882
          - no rows returned for version select:
            github.com/coder/coder/v2/cli.ConnectToPostgres
                /Users/runner/work/coder/coder/cli/server.go:2363
    clitest.go:172: 
        	Error Trace:	/Users/runner/work/coder/coder/cli/clitest/clitest.go:172
        	            				/Users/runner/work/_tool/go/1.24.2/arm64/src/runtime/asm_arm64.s:1223
        	Error:      	Received unexpected error:
        	            	running command "coder server": connect to postgres:
        	            	    github.com/coder/coder/v2/cli.(*RootCmd).Server.func2
        	            	        /Users/runner/work/coder/coder/cli/server.go:754
        	            	  - connect to postgres:
        	            	    github.com/coder/coder/v2/cli.getAndMigratePostgresDB
        	            	        /Users/runner/work/coder/coder/cli/server.go:2882
        	            	  - no rows returned for version select:
        	            	    github.com/coder/coder/v2/cli.ConnectToPostgres
        	            	        /Users/runner/work/coder/coder/cli/server.go:2363
        	Test:       	TestServer/Logging/CreatesFile
        --- FAIL: TestServer/Logging/CreatesFile (2.82s)
```

--- a/cli/clitest/clitest.go
+++ b/cli/clitest/clitest.go
@@ -168,6 +168,12 @@ func StartWithAssert(t *testing.T, inv *serpent.Invocation, assertCallback func(
 		switch {
 		case errors.Is(err, context.Canceled):
 			return
+		case err != nil && strings.Contains(err.Error(), "driver: bad connection"):
+			// When we cancel the context on a query that's being executed within
+			// a transaction, sometimes, instead of a context.Canceled error we get
+			// a "driver: bad connection" error.
+			// https://github.com/lib/pq/issues/1137
+			return
 		default:
 			assert.NoError(t, err)
 		}

--- a/cli/server.go
+++ b/cli/server.go
@@ -2359,6 +2359,10 @@ func ConnectToPostgres(ctx context.Context, logger slog.Logger, driver string, d
 	if err != nil {
 		return nil, xerrors.Errorf("get postgres version: %w", err)
 	}
+	defer version.Close()
+	if version.Err() != nil {
+		return nil, xerrors.Errorf("version select: %w", version.Err())
+	}
 	if !version.Next() {
 		return nil, xerrors.Errorf("no rows returned for version select")
 	}
@@ -2367,7 +2371,6 @@ func ConnectToPostgres(ctx context.Context, logger slog.Logger, driver string, d
 	if err != nil {
 		return nil, xerrors.Errorf("scan version: %w", err)
 	}
-	_ = version.Close()
 
 	if versionNum < 130000 {
 		return nil, xerrors.Errorf("PostgreSQL version must be v13.0.0 or higher! Got: %d", versionNum)

--- a/coderd/database/pubsub/pubsub.go
+++ b/coderd/database/pubsub/pubsub.go
@@ -552,15 +552,14 @@ func (p *PGPubsub) startListener(ctx context.Context, connectURL string) error {
 			sentErrCh = true
 		}),
 	}
-	select {
-	case err = <-errCh:
-		if err != nil {
-			_ = p.pgListener.Close()
-			return xerrors.Errorf("create pq listener: %w", err)
-		}
-	case <-ctx.Done():
+	// We don't respect context cancellation here. There's a bug in the pq library
+	// where if you close the listener before or while the connection is being
+	// established, the connection will be established anyway, and will not be
+	// closed.
+	// https://github.com/lib/pq/issues/1192
+	if err := <-errCh; err != nil {
 		_ = p.pgListener.Close()
-		return ctx.Err()
+		return xerrors.Errorf("create pq listener: %w", err)
 	}
 	return nil
 }

